pyuvdata/uvdata/uvdata.py

```diff
@@ -5563,14 +5570,179 @@ def compress_by_redundancy(self, tol=1.0, inplace=True, keep_all_metadata=True):
             if inplace is False, return the compressed UVData object
 
         """
+        allowed_methods = ["select", "average"]
+        if method not in allowed_methods:
+            raise ValueError(f"method must be one of {allowed_methods}")
+
         red_gps, centers, lengths, conjugates = self.get_redundancies(
             tol, include_conjugates=True
         )
-
         bl_ants = [self.baseline_to_antnums(gp[0]) for gp in red_gps]
-        return self.select(
-            bls=bl_ants, inplace=inplace, keep_all_metadata=keep_all_metadata
-        )
+
+        if method == "average":
+            # do a metadata only select to get all the metadata right
+            new_obj = self.copy(metadata_only=True)
+            new_obj.select(bls=bl_ants, keep_all_metadata=keep_all_metadata)
+
+            temp_time_array = np.zeros_like(new_obj.time_array)
+            if not self.metadata_only:
+                # initalize the data like arrays
+                temp_data_array = np.zeros(
+                    (new_obj.Nblts, new_obj.Nspws, new_obj.Nfreqs, new_obj.Npols),
+                    dtype=self.data_array.dtype,
+                )
+                temp_nsample_array = np.zeros(
+                    (new_obj.Nblts, new_obj.Nspws, new_obj.Nfreqs, new_obj.Npols),
+                    dtype=self.nsample_array.dtype,
+                )
+                temp_flag_array = np.zeros(
+                    (new_obj.Nblts, new_obj.Nspws, new_obj.Nfreqs, new_obj.Npols),
+                    dtype=self.flag_array.dtype,
+                )
+            for grp_ind, group in enumerate(red_gps):
+                if len(conjugates) > 0:
+                    conj_group = set(group).intersection(conjugates)
+                    reg_group = list(set(group) - conj_group)
+                    conj_group = list(conj_group)
+                else:
+                    reg_group = group
+                    conj_group = []
+                group_times = []
+                group_inds = []
+                conj_group_inds = []
+                conj_group_times = []
+                for bl in reg_group:
+                    bl_inds = np.where(self.baseline_array == bl)[0]
+                    group_inds.extend(bl_inds)
+                    group_times.extend(self.time_array[bl_inds])
+                for bl in conj_group:
+                    bl_inds = np.where(self.baseline_array == bl)[0]
+                    conj_group_inds.extend(bl_inds)
+                    conj_group_times.extend(self.time_array[bl_inds])
```

Comment:
We might be able to do this part without having to loop through every baseline with a `numpy.searchsorted` kind of like the uvw calculation. However, it would require the object to be in the `baseline,time` ordering to ensure that the `baseline_array` is sorted with baseline number increasing. But I'm not sure it would really save us too much. Probably worth some profiling to investigate later?

Reply:
yeah, lets do that.

Reply:
should we make an issue?

Reply:
I've made one so we can come back to this (#863 )